Obtaining project build output files failed when using finalName

**Environment**:

- *Jib version:* 2.4.0
- *Build tool:* maven:3.6.3-amazoncorretto-11
- *OS:* Amazon Linux 2

**Description of the issue**:
When spring-boot-maven-plugin finalName is used jib cannot find project build output file.


**Steps to reproduce**:

    mvn package jib:build ...


**`jib-maven-plugin` Configuration**: <!-- Delete this section if not used -->
```xml
    <plugin>
         <groupId>org.springframework.boot</groupId>
         <artifactId>spring-boot-maven-plugin</artifactId>
         <configuration>
             <finalName>${project.artifactId}</finalName>
          </configuration>
    </plugin>

    <plugin>
           <groupId>com.google.cloud.tools</groupId>
           <artifactId>jib-maven-plugin</artifactId>
           <version>2.4.0</version>
           <configuration>
                 ....
                 <containerizingMode>packaged</containerizingMode>
            </configuration>
    </plugin>
```

**Log output**:

[ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:2.4.0:build (default-cli) on project registry: Obtaining project build output files failed; make sure you have packaged your project before trying to build the image. (Did you accidentally run "mvn clean jib:build" instead of "mvn clean package jib:build"?): /home/builder/src/myapp/target/myapp-3.0.2-SNAPSHOT.jar.original -> [Help 1]

--- a/jib-maven-plugin/CHANGELOG.md
+++ b/jib-maven-plugin/CHANGELOG.md
@@ -16,6 +16,7 @@ All notable changes to this project will be documented in this file.
 - Fixed `NullPointerException` when the `"auths":` section in `~/.docker/config.json` has an entry with no `"auth":` field. ([#2535](https://github.com/GoogleContainerTools/jib/issues/2535))
 - Fixed `NullPointerException` to return a helpful message when a server does not provide any message in certain error cases (400 Bad Request, 404 Not Found, and 405 Method Not Allowed). ([#2532](https://github.com/GoogleContainerTools/jib/issues/2532))
 - Now supports sending client certificate (for example, via the `javax.net.ssl.keyStore` and `javax.net.ssl.keyStorePassword` system properties) and thus enabling mutual TLS authentication. ([#2585](https://github.com/GoogleContainerTools/jib/issues/2585), [#2226](https://github.com/GoogleContainerTools/jib/issues/2226))
+- Fixed build failure with `<containerizingMode>packaged` in Spring Boot projects where Jib assumed a wrong JAR path when `<finalName>` or `<classifier>` is configured in Spring Boot. ([#2565](https://github.com/GoogleContainerTools/jib/issues/2565))
 
 ## 2.4.0
 

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
@@ -517,16 +517,23 @@ public class MavenProjectProperties implements ProjectProperties {
       }
     }
 
+    String finalName = project.getBuild().getFinalName();
     String suffix = ".jar";
-    if (jarRepackagedBySpringBoot()) {
+
+    Optional<Xpp3Dom> bootConfiguration = getSpringBootRepackageConfiguration();
+    if (bootConfiguration.isPresent()) {
       log(LogEvent.lifecycle("Spring Boot repackaging (fat JAR) detected; using the original JAR"));
-      if (outputDirectory.equals(buildDirectory)) { // Spring renames original only when needed
-        suffix += ".original";
+
+      // Spring renames original JAR only when replacing it, so check if the names are clashing.
+      if (outputDirectory.equals(buildDirectory)) {
+        Optional<String> bootFinalName = getChildValue(bootConfiguration.get(), "finalName");
+        if (!bootFinalName.isPresent() || bootFinalName.get().equals(finalName)) {
+          suffix = ".jar.original";
+        }
       }
     }
 
-    String noSuffixJarName =
-        project.getBuild().getFinalName() + (classifier == null ? "" : '-' + classifier);
+    String noSuffixJarName = finalName + (classifier == null ? "" : '-' + classifier);
     Path jarPath = outputDirectory.resolve(noSuffixJarName + suffix);
     log(LogEvent.debug("Using JAR: " + jarPath));
 
@@ -542,20 +549,25 @@ public class MavenProjectProperties implements ProjectProperties {
     return newJarPath;
   }
 
+  /**
+   * Returns Spring Boot {@code &lt;configuration&gt;} if the Spring Boot plugin is configured to
+   * run the {@code repackage} goal to create a Spring Boot artifact.
+   */
   @VisibleForTesting
-  boolean jarRepackagedBySpringBoot() {
+  Optional<Xpp3Dom> getSpringBootRepackageConfiguration() {
     Plugin springBootPlugin =
         project.getPlugin("org.springframework.boot:spring-boot-maven-plugin");
     if (springBootPlugin != null) {
       for (PluginExecution execution : springBootPlugin.getExecutions()) {
         if (execution.getGoals().contains("repackage")) {
-          Optional<String> skip = getChildValue((Xpp3Dom) execution.getConfiguration(), "skip");
-          boolean skipped = "true".equals(skip.orElse("false"));
-          return !skipped;
+          Xpp3Dom configuration = (Xpp3Dom) execution.getConfiguration();
+
+          boolean skip = Boolean.valueOf(getChildValue(configuration, "skip").orElse("false"));
+          return skip ? Optional.empty() : Optional.of(configuration);
         }
       }
     }
-    return false;
+    return Optional.empty();
   }
 
   @Override

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
@@ -497,7 +497,7 @@ public class MavenProjectProperties implements ProjectProperties {
    */
   @VisibleForTesting
   Path getJarArtifact() throws IOException {
-    String classifier = null;
+    Optional<String> classifier = Optional.empty();
     Path buildDirectory = Paths.get(project.getBuild().getDirectory());
     Path outputDirectory = buildDirectory;
 
@@ -507,7 +507,7 @@ public class MavenProjectProperties implements ProjectProperties {
       for (PluginExecution execution : jarPlugin.getExecutions()) {
         if ("default-jar".equals(execution.getId())) {
           Xpp3Dom configuration = (Xpp3Dom) execution.getConfiguration();
-          classifier = getChildValue(configuration, "classifier").orElse(null);
+          classifier = getChildValue(configuration, "classifier");
           Optional<String> directoryString = getChildValue(configuration, "outputDirectory");
 
           if (directoryString.isPresent()) {
@@ -524,16 +524,20 @@ public class MavenProjectProperties implements ProjectProperties {
     if (bootConfiguration.isPresent()) {
       log(LogEvent.lifecycle("Spring Boot repackaging (fat JAR) detected; using the original JAR"));
 
-      // Spring renames original JAR only when replacing it, so check if the names are clashing.
-      if (outputDirectory.equals(buildDirectory)) {
-        Optional<String> bootFinalName = getChildValue(bootConfiguration.get(), "finalName");
-        if (!bootFinalName.isPresent() || bootFinalName.get().equals(finalName)) {
-          suffix = ".jar.original";
-        }
+      // Spring renames original JAR only when replacing it, so check if the paths are clashing.
+      Optional<String> bootFinalName = getChildValue(bootConfiguration.get(), "finalName");
+      Optional<String> bootClassifier = getChildValue(bootConfiguration.get(), "classifier");
+
+      boolean sameDirectory = outputDirectory.equals(buildDirectory);
+      // If Boot <finalName> is undefined, it uses the default project <finalName>.
+      boolean sameFinalName = !bootFinalName.isPresent() || finalName.equals(bootFinalName.get());
+      boolean sameClassifier = classifier.equals(bootClassifier);
+      if (sameDirectory && sameFinalName && sameClassifier) {
+        suffix = ".jar.original";
       }
     }
 
-    String noSuffixJarName = finalName + (classifier == null ? "" : '-' + classifier);
+    String noSuffixJarName = finalName + (classifier.isPresent() ? '-' + classifier.get() : "");
     Path jarPath = outputDirectory.resolve(noSuffixJarName + suffix);
     log(LogEvent.debug("Using JAR: " + jarPath));
 

--- a/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
+++ b/jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java
@@ -513,6 +513,7 @@ public class MavenProjectProperties implements ProjectProperties {
           if (directoryString.isPresent()) {
             outputDirectory = project.getBasedir().toPath().resolve(directoryString.get());
           }
+          break;
         }
       }
     }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
@@ -636,25 +636,25 @@ public class MavenProjectPropertiesTest {
   }
 
   @Test
-  public void testIsWarProject_WarPackagingIsWar() {
+  public void testIsWarProject_warPackagingIsWar() {
     Mockito.when(mockMavenProject.getPackaging()).thenReturn("war");
     Assert.assertTrue(mavenProjectProperties.isWarProject());
   }
 
   @Test
-  public void testIsWarProject_GwtAppPackagingIsWar() {
+  public void testIsWarProject_gwtAppPackagingIsWar() {
     Mockito.when(mockMavenProject.getPackaging()).thenReturn("gwt-app");
     Assert.assertTrue(mavenProjectProperties.isWarProject());
   }
 
   @Test
-  public void testIsWarProject_JarPackagingIsNotWar() {
+  public void testIsWarProject_jarPackagingIsNotWar() {
     Mockito.when(mockMavenProject.getPackaging()).thenReturn("jar");
     Assert.assertFalse(mavenProjectProperties.isWarProject());
   }
 
   @Test
-  public void testIsWarProject_GwtLibPackagingIsNotWar() {
+  public void testIsWarProject_gwtLibPackagingIsNotWar() {
     Mockito.when(mockMavenProject.getPackaging()).thenReturn("gwt-lib");
     Assert.assertFalse(mavenProjectProperties.isWarProject());
   }
@@ -983,6 +983,32 @@ public class MavenProjectPropertiesTest {
   }
 
   @Test
+  public void testGetJarArtifact_originalJarIfSpringBoot_differentFinalNames() throws IOException {
+    Path buildDirectory = temporaryFolder.newFolder("target").toPath();
+    Files.createFile(buildDirectory.resolve("helloworld-1.jar"));
+    Mockito.when(mockMavenProject.getBasedir()).thenReturn(temporaryFolder.getRoot());
+    Mockito.when(mockBuild.getDirectory()).thenReturn(buildDirectory.toString());
+    Mockito.when(mockBuild.getFinalName()).thenReturn("helloworld-1");
+
+    Mockito.when(mockMavenProject.getPlugin("org.apache.maven.plugins:maven-jar-plugin"))
+        .thenReturn(mockPlugin);
+    Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
+    Mockito.when(mockPluginExecution.getId()).thenReturn("default-jar");
+    Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
+    addXpp3DomChild(pluginConfiguration, "outputDirectory", "target");
+
+    Xpp3Dom bootPluginConfiguration = setUpSpringBootFatJar();
+    addXpp3DomChild(bootPluginConfiguration, "finalName", "boot-helloworld-1");
+
+    Assert.assertEquals(
+        buildDirectory.resolve("helloworld-1.jar"), mavenProjectProperties.getJarArtifact());
+
+    mavenProjectProperties.waitForLoggingThread();
+    Mockito.verify(mockLog)
+        .info("Spring Boot repackaging (fat JAR) detected; using the original JAR");
+  }
+
+  @Test
   public void testGetWarArtifact() {
     Mockito.when(mockBuild.getDirectory()).thenReturn(Paths.get("/foo/bar").toString());
     Mockito.when(mockBuild.getFinalName()).thenReturn("helloworld-1");
@@ -1063,7 +1089,8 @@ public class MavenProjectPropertiesTest {
     return unzipTarget;
   }
 
-  private void setUpSpringBootFatJar() {
+  private Xpp3Dom setUpSpringBootFatJar() {
+    Xpp3Dom pluginConfiguration = new Xpp3Dom("configuration");
     PluginExecution execution = Mockito.mock(PluginExecution.class);
     Plugin plugin = Mockito.mock(Plugin.class);
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
@@ -1071,5 +1098,6 @@ public class MavenProjectPropertiesTest {
     Mockito.when(plugin.getExecutions()).thenReturn(Arrays.asList(execution));
     Mockito.when(execution.getGoals()).thenReturn(Arrays.asList("repackage"));
     Mockito.when(execution.getConfiguration()).thenReturn(pluginConfiguration);
+    return pluginConfiguration;
   }
 }

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
@@ -955,6 +955,58 @@ public class MavenProjectPropertiesTest {
   }
 
   @Test
+  public void testGetJarArtifact_originalJarIfSpringBoot_differentFinalNames() throws IOException {
+    Path buildDirectory = temporaryFolder.newFolder("target").toPath();
+    Files.createFile(buildDirectory.resolve("helloworld-1.jar"));
+    Mockito.when(mockMavenProject.getBasedir()).thenReturn(temporaryFolder.getRoot());
+    Mockito.when(mockBuild.getDirectory()).thenReturn(buildDirectory.toString());
+    Mockito.when(mockBuild.getFinalName()).thenReturn("helloworld-1");
+
+    Mockito.when(mockMavenProject.getPlugin("org.apache.maven.plugins:maven-jar-plugin"))
+        .thenReturn(mockPlugin);
+    Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
+    Mockito.when(mockPluginExecution.getId()).thenReturn("default-jar");
+    Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
+    addXpp3DomChild(pluginConfiguration, "outputDirectory", "target");
+
+    Xpp3Dom bootPluginConfiguration = setUpSpringBootFatJar();
+    addXpp3DomChild(bootPluginConfiguration, "finalName", "boot-helloworld-1");
+
+    Assert.assertEquals(
+        buildDirectory.resolve("helloworld-1.jar"), mavenProjectProperties.getJarArtifact());
+
+    mavenProjectProperties.waitForLoggingThread();
+    Mockito.verify(mockLog)
+        .info("Spring Boot repackaging (fat JAR) detected; using the original JAR");
+  }
+
+  @Test
+  public void testGetJarArtifact_originalJarIfSpringBoot_differentClassifier() throws IOException {
+    Path buildDirectory = temporaryFolder.newFolder("target").toPath();
+    Files.createFile(buildDirectory.resolve("helloworld-1.jar"));
+    Mockito.when(mockMavenProject.getBasedir()).thenReturn(temporaryFolder.getRoot());
+    Mockito.when(mockBuild.getDirectory()).thenReturn(buildDirectory.toString());
+    Mockito.when(mockBuild.getFinalName()).thenReturn("helloworld-1");
+
+    Mockito.when(mockMavenProject.getPlugin("org.apache.maven.plugins:maven-jar-plugin"))
+        .thenReturn(mockPlugin);
+    Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
+    Mockito.when(mockPluginExecution.getId()).thenReturn("default-jar");
+    Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
+    addXpp3DomChild(pluginConfiguration, "outputDirectory", "target");
+
+    Xpp3Dom bootPluginConfiguration = setUpSpringBootFatJar();
+    addXpp3DomChild(bootPluginConfiguration, "classifier", "boot-class");
+
+    Assert.assertEquals(
+        buildDirectory.resolve("helloworld-1.jar"), mavenProjectProperties.getJarArtifact());
+
+    mavenProjectProperties.waitForLoggingThread();
+    Mockito.verify(mockLog)
+        .info("Spring Boot repackaging (fat JAR) detected; using the original JAR");
+  }
+
+  @Test
   public void testGetJarArtifact_originalJarCopiedIfSpringBoot_sameDirectory() throws IOException {
     Path buildDirectory = temporaryFolder.newFolder("target").toPath();
     Files.createFile(buildDirectory.resolve("helloworld-1.jar.original"));
@@ -976,32 +1028,6 @@ public class MavenProjectPropertiesTest {
     Assert.assertEquals(
         tempDirectory.resolve("helloworld-1.original.jar"),
         mavenProjectProperties.getJarArtifact());
-
-    mavenProjectProperties.waitForLoggingThread();
-    Mockito.verify(mockLog)
-        .info("Spring Boot repackaging (fat JAR) detected; using the original JAR");
-  }
-
-  @Test
-  public void testGetJarArtifact_originalJarIfSpringBoot_differentFinalNames() throws IOException {
-    Path buildDirectory = temporaryFolder.newFolder("target").toPath();
-    Files.createFile(buildDirectory.resolve("helloworld-1.jar"));
-    Mockito.when(mockMavenProject.getBasedir()).thenReturn(temporaryFolder.getRoot());
-    Mockito.when(mockBuild.getDirectory()).thenReturn(buildDirectory.toString());
-    Mockito.when(mockBuild.getFinalName()).thenReturn("helloworld-1");
-
-    Mockito.when(mockMavenProject.getPlugin("org.apache.maven.plugins:maven-jar-plugin"))
-        .thenReturn(mockPlugin);
-    Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
-    Mockito.when(mockPluginExecution.getId()).thenReturn("default-jar");
-    Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
-    addXpp3DomChild(pluginConfiguration, "outputDirectory", "target");
-
-    Xpp3Dom bootPluginConfiguration = setUpSpringBootFatJar();
-    addXpp3DomChild(bootPluginConfiguration, "finalName", "boot-helloworld-1");
-
-    Assert.assertEquals(
-        buildDirectory.resolve("helloworld-1.jar"), mavenProjectProperties.getJarArtifact());
 
     mavenProjectProperties.waitForLoggingThread();
     Mockito.verify(mockLog)

--- a/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
+++ b/jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesTest.java
@@ -775,56 +775,65 @@ public class MavenProjectPropertiesTest {
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_pluginNotApplied() {
-    Assert.assertFalse(mavenProjectProperties.jarRepackagedBySpringBoot());
+  public void testGetSpringBootRepackageConfiguration_pluginNotApplied() {
+    Assert.assertEquals(
+        Optional.empty(), mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_noExecutions() {
+  public void testGetSpringBootRepackageConfiguration_noExecutions() {
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
         .thenReturn(mockPlugin);
     Mockito.when(mockPlugin.getExecutions()).thenReturn(Collections.emptyList());
-    Assert.assertFalse(mavenProjectProperties.jarRepackagedBySpringBoot());
+    Assert.assertEquals(
+        Optional.empty(), mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_noRepackageGoal() {
+  public void testGetSpringBootRepackageConfiguration_noRepackageGoal() {
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
         .thenReturn(mockPlugin);
     Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
     Mockito.when(mockPluginExecution.getGoals()).thenReturn(Arrays.asList("goal", "foo", "bar"));
-    Assert.assertFalse(mavenProjectProperties.jarRepackagedBySpringBoot());
+    Assert.assertEquals(
+        Optional.empty(), mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_repackageGoal() {
+  public void testGetSpringBootRepackageConfiguration_repackageGoal() {
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
         .thenReturn(mockPlugin);
     Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
     Mockito.when(mockPluginExecution.getGoals()).thenReturn(Arrays.asList("goal", "repackage"));
-    Assert.assertTrue(mavenProjectProperties.jarRepackagedBySpringBoot());
+    Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
+    Assert.assertEquals(
+        Optional.of(pluginConfiguration),
+        mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_skipped() {
+  public void testGetSpringBootRepackageConfiguration_skipped() {
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
         .thenReturn(mockPlugin);
     Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
     Mockito.when(mockPluginExecution.getGoals()).thenReturn(Arrays.asList("repackage"));
     Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
     addXpp3DomChild(pluginConfiguration, "skip", "true");
-    Assert.assertFalse(mavenProjectProperties.jarRepackagedBySpringBoot());
+    Assert.assertEquals(
+        Optional.empty(), mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
-  public void testJarRepackagedBySpringBoot_skipNotTrue() {
+  public void testGetSpringBootRepackageConfiguration_skipNotTrue() {
     Mockito.when(mockMavenProject.getPlugin("org.springframework.boot:spring-boot-maven-plugin"))
         .thenReturn(mockPlugin);
     Mockito.when(mockPlugin.getExecutions()).thenReturn(Arrays.asList(mockPluginExecution));
     Mockito.when(mockPluginExecution.getGoals()).thenReturn(Arrays.asList("repackage"));
     Mockito.when(mockPluginExecution.getConfiguration()).thenReturn(pluginConfiguration);
     addXpp3DomChild(pluginConfiguration, "skip", null);
-    Assert.assertTrue(mavenProjectProperties.jarRepackagedBySpringBoot());
+    Assert.assertEquals(
+        Optional.of(pluginConfiguration),
+        mavenProjectProperties.getSpringBootRepackageConfiguration());
   }
 
   @Test
@@ -1061,5 +1070,6 @@ public class MavenProjectPropertiesTest {
         .thenReturn(plugin);
     Mockito.when(plugin.getExecutions()).thenReturn(Arrays.asList(execution));
     Mockito.when(execution.getGoals()).thenReturn(Arrays.asList("repackage"));
+    Mockito.when(execution.getConfiguration()).thenReturn(pluginConfiguration);
   }
 }

